proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1051 @@
+
+A notable exception is the defunctionalisation arrow from the
+`singletons <https://hackage.haskell.org/package/singletons-2.7/docs/Data-Singletons.html#t:-126--62->`_ library::
+
+  type (~>) a b = TyFun a b -> Type
+
+which really refers to a kind-level matchable arrow. However, we expect many
+such use cases to be subsumed by first class higher-order functions introduced
+by this proposal.
+
+Arity of type families
+~~~~~~~~~~~~~~~~~~~~~~
+
+A technical consequence of the proposed framework is that the `arity
+<https://downloads.haskell.org/~ghc/8.10.2/docs/html/users_guide/glasgow_exts.html?highlight=typefamilies#type-family-declarations>`_
+of a type family can not be directly deduced from its kind (although even before
+this proposal that was already the case).
+
+Consider the following two type families ::
+
+  type family Foo (a :: Type) :: Type
+  type family Bar :: Type -> @U Type
+
+Both have the same kind, namely ``Type -> @U Type``, but  the arity of ``Foo``
+is 1, whereas ``Bar`` is nullary. Since partial application is now possible, the
+arities no longer play such an important role. The main place where they still
+show up is in the definitions of type families. Type family equations must bind
+all of their arguments on the left-hand side ::
+
+  type family Foo (a :: Type) :: Type where
+    Foo Int  = Bool
+    Foo Char = Int
+
+but ``Bar``, a nullary type family, can only be defined without arguments and a
+type family on its RHS ::
+
+  type family Bar :: Type -> @U Type where
+    Bar = Foo
+
+Thus the following definition is invalid ::
+
+  type family Bad :: Type -> @U Type where
+    Bad x = Foo x
+
+This is because type families can only be reduced when they are fully
+saturated.
+
+Thus, the relationship between the arity and the kind can be summarised as follows:
+If a type family's arity is ``n``, then its kind will have *at least* its first
+``n`` arrows unmatchable.
+
+Apartness of type families
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+GHC has a notion of "apartness": when matching a type family equation, it needs
+to make sure that no previous equations match or when matching a type class
+instance, that no other instances match. Two types don't match when they are
+apart. The apartness check has three possible outcomes: two types can "unify",
+be "surely apart", or be "maybe apart".
+
+We propose that type families unify with themselves, but they are "maybe apart"
+from every other type. The consequence of this is that writing type families in
+positions where apartness is considered is illegal. That is, both
+
+::
+
+  instance C Id where ...
+
+   type instance F Id = ...
+
+are illegal when ``Id`` is a type family.
+
+Proposed Change Specification
+-----------------------------
+
+The following sections describe a new GHC extension, which can be
+enabled with the pragma ``{-# LANGUAGE UnsaturatedTypeFamilies #-}``.
+The pragma implies ``TypeFamilies``.
```

Comment:
`UnsaturatedTypeFamilies` does not actually guard the ability to partially apply type families, does it? The actual change specification seems to be all about supporting the matchability annotations, in which case it seems something like `MatchabilityAnnotations` would be a better name for the extension.

Reply:
It does guard the ability to partially apply them, and a partial application of a type family suggests turning on `UnsaturatedTypeFamilies`. That way when users partially apply a type family they won't immediately start seeing errors about matchabilities. On that note, it is not part of the proposal but my intention is to present more legible error messages than just "matchabilities failed to unify".

Reply:
Aha, but the "Proposed Change Specification" says nothing about allowing partial application of type families, that should be first bullet point!

I do wonder though if we should allow partial application by default. We tend to require extensions at definition sites rather than use sites, e.g. I don't need to enable `TypeFamilies` to apply a type family or `PatternSynonyms` to match one. On the other hand, your concern about the quality of error messages is valid, and this will be an experimental extension.. So maybe the best thing is to have this extension guard partial application for now, and reconsider allowing partial application by default when we're ready to lift the experimental status.

Reply:
Oops, what a glaring omission! Indeed the proposal didn't mention the most important part :-) fixed!